hearth/lib/hearth/middleware/retry.rb

```diff
@@ -5,49 +5,111 @@ module Middleware
+      #   of attempts that will be made for a single request, including the
+      #   initial attempt.
+      # @param [Boolean] adaptive_retry_wait_to_fill Used only in `adaptive`
+      #   retry mode. When true, the request will sleep until there is
+      #   sufficient client side capacity to retry the request. When false,
+      #   the request will raise a `CapacityNotAvailableError` and will not
+      #   retry instead of sleeping.
+      def initialize(app, retry_mode:, max_attempts:,
+                     adaptive_retry_wait_to_fill:, error_inspector_class:,
+                     retry_quota:, client_rate_limiter:)
         @app = app
+
+        # public config
+        @retry_mode = retry_mode
         @max_attempts = max_attempts
-        @max_delay = max_delay
+        @adaptive_retry_wait_to_fill = adaptive_retry_wait_to_fill
+
+        # undocumented options
+        @error_inspector_class = error_inspector_class
+        @retry_quota = retry_quota
+        @client_rate_limiter = client_rate_limiter
+
+        # instance state
+        @capacity_amount = nil
+        @retries = 0
       end
 
       # @param input
       # @param context
       # @return [Output]
-      # rubocop:disable Metrics/MethodLength
+      # rubocop:disable Metrics/AbcSize
       def call(input, context)
-        attempt = 1
-        begin
-          output = @app.call(input, context)
-          raise output.error if output.error
-
-          output
-        rescue Hearth::ApiError => e
-          return output if !e.retryable? || attempt >= @max_attempts
-
-          Kernel.sleep(backoff_with_jitter(attempt))
-          attempt += 1
-          retry
-        rescue Hearth::HTTP::NetworkingError => e
-          raise e if attempt >= @max_attempts
-
-          Kernel.sleep(backoff_with_jitter(attempt))
-          attempt += 1
-          retry
-        end
+        acquire_token
+        output = @app.call(input, context)
+        error_inspector = @error_inspector_class.new(
+          output.error, context.response.status
+        )
+        request_bookkeeping(output, error_inspector)
+
+        return output unless retryable?(context, output, error_inspector)
+
+        return output if @retries >= @max_attempts - 1
+
+        @capacity_amount = @retry_quota.checkout_capacity(error_inspector)
+        return output unless @capacity_amount.positive?
+
+        delay = [Kernel.rand * (2**@retries), MAX_BACKOFF].min
+        Kernel.sleep(delay)
+        retry_request(input, context, output)
       end
-      # rubocop:enable Metrics/MethodLength
+      # rubocop:enable Metrics/AbcSize
 
       private
 
-      # https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
-      def backoff_with_jitter(attempt)
-        # scales like 1,2,4,8
-        Kernel.rand * [@max_delay, 2**(attempt - 1)].min
+      def acquire_token
+        return unless @retry_mode != 'adaptive'
```

Comment:
unless != is a bit confusing here.  Is this logic even correct?

Reply:
Yeah I think that's correct. It's a no op for standard.